---
title: "BOO 2025 - Example Analysis"
subtitle: "Script 6: Functional enrichment - Answers"
date: "`r Sys.Date()`" 
author: 
  Student name
output:
  html_document:
    code_download: true
    theme: united
    lowlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

***

> In this script, you will biologically interpret your differentially expressed genes.

***

# Setup

## Clean

As before, we perform several setup steps at the start of the script to ensure our work is reproducible and clear. 

**Exercise 1: Download the R markdown template, clean your environment, and set the folhighing variables:**

* `root_dir` - project folder
* `low_store` - location of DEG results for the low toxicity compound
* `high_store` - location of DEG results for the high toxicity compound

```{r clean}
 rm(list = ls())


low_store <- file.path("Output/low_store.csv")
high_store <- file.path("Output/high_store.csv")

```

***

## Packages

In this script, we introduce a new package: `enrichR`, a package for gene set enrichment analysis that identifies biological pathways, gene ontology terms, and other functional annotations significantly associated with a given list of genes.

**Exercise 2: Load `enrichR` and `tidyverse` into your R environment:**

```{r load-packages, warning=F, message=F}
library(tidyverse)
library(enrichR)
```

***

## Load data

**Exercise 3: Load the DEG results R environment:**

```{r load-data, warning=FALSE, message=FALSE}
 rm(list = ls())

res_low  <- read_csv("output/low_store.csv")

res_high  <- read_csv("output/high_store.csv")


```

***

# ORA

## Databases

When performing any type of gene set analysis, you need to specify the gene sets of interest. This is one of the pitfalls but also benefits of this type of analysis.

**Exercise 4: Use `listEnrichrDbs()` and look at the gene set databases available. Select a few of interest and describe them:**

<details>
  <summary><strong>Hint</strong></summary>

  Sometimes a specific database has different version for different years. Try to select the most recent ones using `grepl`.

</details>

```{r}
listEnrichrDbs() %>% 
  filter(grepl("202", libraryName))
```

* `MSigDB_Hallmark_2020`: gene sets curated by the Broad Institute from the Molecular Signatures Database (MSigDB), repesenting coherent biological states or processes aimed to reduce redundancy in gene set databases.
* `TG_GATES_2020`: derived from the Toxicogenomics Project-Genomics Assisted Toxicity Evaluation System (TG-GATEs), this database links gene expression changes to specific chemical exposures in rodent and human hepatocytes, providing mechanistic insights into toxicity and drug safety.
* `GO_Biological_Process_2025`, `GO_Cellular_Component_2025`, and `GO_Molecular_Function_2025` are part of the Gene Ontology (GO) Consortium, and capture structured biological processes (e.g. inflammatory response), subcellular locations (e.g. mitochondrial matrix), and molecular activities (e.g. ATP binding), respectively.
* `WikiPathways_2024_Human` is an open-source, community-curated collection of human biological pathways hosted by the WikiPathways project, aimed at enhancing transparency and collaboration in pathway annotation across a range of processes and diseases. 
* `Reactome_Pathways_2024` is a peer-reviewed and manually curated database of human pathways developed by the Reactome Consortium, emphasizing detailed molecular interactions and regulatory mechanisms across signalling, metabolism, and disease.
* `KEGG_2021_Human` is developed by the Kyoto Encyclopedia of Genes and Genomes (KEGG) and contains human pathway maps integrating genomic, chemical, and functional data to support systems-level understanding of cellular and organismal functions.

***

## DEG List

In order to run ORA you need a list of genes of interest (your DEGs).

**Exercise 5: Make a DEG list for each of the two compounds. How many DEGs do you have for each?**

<details>
  <summary><strong>Hint</strong></summary>

  If you compound shows a dose response, you may want to select genes at a particular concentration of interest.

</details>




```{r}
deg_high <- unique((res_high %>% 
               filter(sig == T))$gene_symbol)

deg_low <- unique((res_low %>% 
               filter(sig == T))$gene_symbol)
```

There are 173 DEGs for the high toxicity compound and 3601 for the low toxicity one.

```{r}
deg_high2 <- sapply(str_split(deg_high, "_",),'[',1) # chang the names from AARS1_3 to AAES1 so the function enrichr can look for it

deg_low2 <- sapply(str_split(deg_low, "_",),'[',1)
```
***

## Enrichr

The `enrichr` function performs over-representation analysis (ORA) using a list of genes in each of the databases specified by the `databases` argument, returning a list.

**Exercise 6: Run `enrichr` using the selected databases and inspect the output for each compound. Are there any pathways you expected?:**

<details>
  <summary><strong>Hint</strong></summary>

  You can use `bind_rows()` to convert lists into data frames and make inspecting the output easier.

</details>

```{r}
ora_high <- enrichr(deg_high2,
                       c("GO_Biological_Process_2025",
                         "GO_Molecular_Function_2025",
                         "Reactome_Pathways_2024",
                         "KEGG_2021_Human"))


```
```{r}
ora_high <- bind_rows(ora_high)

ora_high %>% 
  filter(Adjusted.P.value <= 0.05)
```


There are 2501 enriched terms in the DEGs from the compound (DIPA). Some of the top terms are expected responses from such an exposure:


* `Retinoid Metabolic Process & Retinoic Acid Metabolic Process`:These terms relate to the metabolism of retinoids (Vitamin A and its derivatives like retinoic acid), which are crucial for vision, growth, cell differentiation, and immune function. The liver plays a central role in storing and metabolizing retinoids. Cytochrome P450 enzymes are involved in this process. Enrichment here suggests your DEGs are involved in altering how Vitamin A is processed or utilized. **Plausible** (due to shared enzyme systems or broad metabolic impact). The liver is central to both retinoid and steroid metabolism. Many enzymes involved (especially CYPs) overlap with xenobiotic metabolism.

* `Steroid Metabolic Process`:This involves the synthesis, modification, and breakdown of steroids, which include hormones like cholesterol, sex hormones, and corticosteroids. The liver is a key organ for steroid metabolism. Enrichment indicates an impact on these hormonal and metabolic pathways.
* `Xenobiotic Catabolic Process`: "Xenobiotic" refers to foreign chemical substances not normally found in the body (like drugs or environmental toxins). "Catabolic process" means breaking them down. The liver is the primary organ for detoxifying and metabolizing xenobiotics, often via cytochrome P450 enzymes. Enrichment here is a strong indicator that the cells are responding to the compound as a foreign substance and are trying to metabolize or eliminate it. This is a very common and expected finding in toxicology studies. **Expected**. DIPA is a foreign chemical (xenobiotic). Liver cells (PHH) are the primary site for metabolizing such compounds. 

* `Aminoglycoside Antibiotic Metabolic Process, Daunorubicin Metabolic Process, Doxorubicin Metabolic Proces`:hese terms refer to the metabolism of specific drugs (aminoglycoside antibiotics, and the chemotherapy drugs daunorubicin and doxorubicin). While your experiment might not use these exact drugs, the enzymes and pathways involved in their metabolism (often cytochrome P450s and other drug-metabolizing enzymes) can overlap with the metabolism of your test compound. Enrichment suggests that the DEGs are part of these general drug metabolism pathways. **Plausible** While DIPA isn't these specific drugs, these terms often highlight the activity of common drug-metabolizing enzyme systems (e.g., cytochrome P450s, conjugating enzymes) that the liver uses for a wide range of xenobiotics. 

* `Polyketide Metabolic Process`: This is generally less directly related to the metabolism of an industrial chemical like DIPA. Its enrichment might point to effects on very broad or less specific enzyme systems, or it could be a weaker, less specific signal.



***

```{r}
ora_low <- enrichr(deg_low2,
                       c("GO_Biological_Process_2025",
                         "GO_Molecular_Function_2025",
                         "Reactome_Pathways_2024",
                         "KEGG_2021_Human"))

ora_low <- bind_rows(ora_low)

ora_low %>% filter(Adjusted.P.value <= 0.05)
```

There are 7953 enriched terms in the DEGs from the compound (tyrosine).
* **Translation** and **Cytoplasmic Translation**: These terms relate to the process of protein synthesis, where genetic information encoded in mRNA is translated into proteins by ribosomes in the cytoplasm. Enrichment here suggests your DEGs are heavily involved in making new proteins. **Plausible**, often as a coordinated cellular response. These terms all relate to the machinery of producing proteins and other large molecules.
 
* **Macromolecule Biosynthetic Process**: This is a broader category that includes the synthesis of large biological molecules like proteins, nucleic acids (DNA/RNA), and polysaccharides. Its enrichment, along with "Translation," points to significant changes in cellular building activities.
* **Gene Expression**: This is a very general term covering the entire process of producing a functional gene product (either RNA or protein) from the information encoded in a gene. Its enrichment indicates widespread changes in how genes are being used.
* **Proton Motive Force-Driven ATP Synthesis** and **Proton Motive Force-Driven Mitochondrial ATP Synthesis ** and **Oxidative Phosphorylation** and **Aerobic Electron Transport Chain** and **Mitochondrial ATP Synthesis Coupled Electron Transport **:These are all closely related terms pointing to mitochondrial energy production. They describe the processes by which cells generate ATP (the main energy currency) through the electron transport chain and oxidative phosphorylation within mitochondria. Enrichment of these terms strongly suggests that the compound treatment is significantly impacting cellular energy metabolism, particularly mitochondrial function.
* **RNA Processing**:  This involves the modification of RNA molecules after transcription, including splicing, capping, and polyadenylation, which are crucial steps in producing mature and functional RNA. Enrichment here indicates that the regulation or processing of RNA is affected by the treatment.
***

## Plots

## Bar plots

Using `geom_col`, we can visualize the results for each of the compounds.

**Exercise 7: Take a look at the top enriched terms for each compound:**

```{r}
ora_high %>% 
  arrange(Adjusted.P.value, P.value) %>% 
  head(10) %>%
  mutate(log10p = -log10(P.value)) %>% 
  separate(Overlap, 
           sep="/", 
           into=c("Genes_found", "Size"),
           remove=F) %>% 
  mutate(Size = as.numeric(Size)) %>% 
  ggplot(aes(y=reorder(Term, log10p))) + 
  geom_col(aes(x=log10p, fill = Size),
           color='black') +
  ylab('') + xlab('-log10p') + ggtitle("Top ten enriched terms:DIPA ") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 30)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=7))

#This bar plot visualizes the top ten most significantly enriched biological terms or pathways for your differentially expressed genes, likely for one of your compounds. The length of each bar represents the statistical significance (-log10 of the P-value, so longer bars are more significant), and the fill color intensity of the bars indicates the size of the gene set (pathway), with darker blues representing larger gene sets.
```

```{r}
ora_low %>% 
  arrange(Adjusted.P.value, P.value) %>% 
  head(10) %>%
  mutate(log10p = -log10(P.value)) %>% 
  separate(Overlap, 
           sep="/", 
           into=c("Genes_found", "Size"),
           remove=F) %>% 
  mutate(Size = as.numeric(Size),
         Genes = str_wrap(Genes, width = 10)) %>% 
  ggplot(aes(y=reorder(Term, log10p))) + 
  geom_col(aes(x=log10p, fill = Size),
           color='black') +
  ylab('') + xlab('-log10p') + ggtitle("Top ten enriched terms:Tyrosine ") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 30)) +
  theme_bw() +
  theme(axis.text.y = element_text(size=7))
```

***

## Upset plot

You may also want to visualize the top enriched terms for both compounds together.

**Exercise: Use `geom_tile()` to do this:**

First we select terms to use in the plot (top 10 for each compound):

```{r}
top_terms <- c((ora_high %>% head(10))$Term, (ora_low %>% head(10))$Term)
top_terms
```

***

Then we bind the results together, making a new variable specifying if the compound is high or low toxicity:

```{r}
ora_high$Toxicity <- "High"
ora_low$Toxicity <- "Low"

ora_both <- rbind(ora_high, ora_low)
ora_both
```

Then, we can make the plot:

```{r}
ora_both %>% 
  filter(Term %in% top_terms) %>% 
  ggplot(aes(y=Term, x=Toxicity)) +
  geom_tile(color="grey50", fill="grey90", linewidth = 0.3) +
  geom_point(aes(fill=Adjusted.P.value, size=Odds.Ratio), shape=21, color='black') +
  ylab("") +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50),
                   expand=c(0,0)) +
  scale_x_discrete(expand=c(0,0)) +
  scale_fill_gradientn(colors = c('dodgerblue', 'black')) +
  ggtitle('Top enriched terms in both compounds:') +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size=6),
    panel.background = element_rect(fill='white', color='grey50'),
    panel.grid = element_blank()
  )
```

We can see that the enrichment of terms is greater (in terms of significance and effect size) in the high toxicity compound. For hydroquinone, strong negative regulation of insulin-like growth factor receptor signaling pathway suggests possible interference with growth or survival signaling, perhaps more expected in a high toxicity compound. For TEG on the other hand, smaller odds ratios overall and less significant p-values are consistent with milder or adaptive cellular responses.

***

# Session Info

```{r session-info}
devtools::session_info()
```

***

